src/librustc_typeck/check/demand.rs

```diff
@@ -374,6 +379,77 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         None
     }
 
+    // This function checks if the specified expression is a built-in range literal
+    // (See: librustc/hir/lowering.rs::LoweringContext::lower_expr() )
+    fn is_range_literal(&self, expr: &hir::Expr) -> bool {
+        use hir::{Path, QPath, ExprKind, TyKind};
+
+        // we support `::std::ops::Range` and `::std::core::Range` prefixes
+        // (via split on "|")
+        let ops_path = ["{{root}}", "std|core", "ops"];
+
+        let is_range_path = |path: &Path| {
```

Comment:
I might be alone in this (cc @estebank), but I think for a short path like this, a more explicit approach (like below) is more readable. (I specifically don't like splitting on "|" because it feels a little too ad hoc.)
```rust
let is_range_path = |path: &Path| {
    let segs = path.segments.iter().map(|seg| seg.ident.as_str());
	if let (Some("{{root}}"), std_core, "ops", range, None) = (segs.next(), segs.next(), segs.next(), segs.next(), segs.next()) {
		(std_core == "std" || std_core == "ops") && range.starts_with("Range")
	} else {
		false
	}
};
```

Reply:
That's a great suggestion! I was looking for something concise like this syntax-wise, it didn't occur to me that I could use `.next()` and rely on it returning `None` when there's no more elements.